text/0000-ques-in-main.md

```diff
@@ -0,0 +1,514 @@
+environment, have to contain extra boilerplate in order to comply with
+those environments' conventions.  A typical construction is
+
+``` rust
+fn inner_main() -> Result<(), ErrorT> {
+    // ... stuff which may fail ...
+    Ok(())
+}
+
+fn main() -> () {
+    use std::process::exit;
+    use std::io::{Write,stderr};
+    use libc::{EXIT_SUCCESS, EXIT_FAILURE};
+
+    exit(match inner_main() {
+        Ok(_) => EXIT_SUCCESS,
+
+        Err(ref err) => {
+            let progname = get_program_name();
+            writeln!(stderr(), "{}: {}\n", progname, err);
+
+            EXIT_FAILURE
+        }
+    })
+}
+```
+
+Both of these problems can be solved at once if the compiler and/or
+libstd are taught to recognize a `main` that returns
+`Result<(), ErrorT>` and supply boilerplate similar to the above.
+
+# Detailed design
+[design]: #detailed-design
+
+The design goals for this new feature are, in decreasing order of
+importance:
+
+1. The `?` operator should be usable in `main` (and `#[test]`
+   functions and doctests).  This entails these functions now
+   returning a richer value than `()`.
+1. Existing code with `fn main() -> ()` should not break.
+1. Errors returned from `main` in a hosted environment should
+   *not* trigger a panic, consistent with the general language
+   principle that panics are only for bugs.
+1. We should take this opportunity to increase consistency with
+   platform conventions for process termination. These often include
+   the ability to pass an "exit status" up to some outer environment,
+   conventions for what that status means, and an expectation that a
+   diagnostic message will be generated when a program fails
+   due to a system error.
+1. We should avoid making life more complicated for people who don't
+   care; on the other hand, if the Easiest Thing is also the Right
+   Thing according to the platform convention, that is better all
+   around.
+
+Goal 1 dictates that the new return type for `main` will be
+`Result<T,E>` for some T and E.  To minimize the necessary changes to
+existing code that wants to start using `?` in `main`, T should be
+_allowed_ to be `()`, but other types in that position may also make
+sense.  The boilerplate shown [above][motivation] will work for any E
+that is `Display`, but that is probably too general for the stdlib; we
+propose only `Error` types should work by default.
+
+[grep]: http://man7.org/linux/man-pages/man1/grep.1.html
+
+## The `Termination` trait
+[the-termination-trait]: #the-termination-trait
+
+When `main` returns a nontrivial value, the runtime needs to know two
+things about it: what error message, if any, to print, and what value
+to pass to `std::process::exit`.  These are naturally encapsulated in
+a trait, which we are tentatively calling `Termination`, with this
+signature:
+
+``` rust
+trait Termination {
+    fn write_diagnostic(&self, progname: &str, stream: &mut Write) -> ();
```

Comment:
I don't know.  I was thinking `std::io::Write` because that's what `set_panic` takes, but depending on how much needs to be in libcore, it could wind up better the other one, I suppose...